Libplanet.MySQLStore/MySQLStore.cs

```diff
@@ -0,0 +1,806 @@
+                        bytes = i.Value;
+                    }
+                }
+            }
+
+            return bytes is null
+                ? 0
+                : MySqlStoreBitConverter.ToInt64(bytes);
+        }
+
+        /// <inheritdoc/>
+        public override IEnumerable<HashDigest<SHA256>> IterateIndexes(
+            Guid chainId,
+            int offset,
+            int? limit)
+        {
+            int count = 0;
+            byte[] prefix = IndexKeyPrefix;
+
+            using QueryFactory db = MySqlUtils.OpenMySqlDB(_connectionString, _compiler);
+            Query query = db.Query(ChainDbName).Select().Where("Prefix", prefix);
+            query = query.Offset(offset);
+            IEnumerable<ChainModel> chains = query.Where("Cf", chainId.ToString()).Get<ChainModel>();
+
+            foreach (var chain in chains)
+            {
+                if (count >= limit)
+                {
+                    break;
+                }
+
+                byte[] value = chain.Value;
+                yield return new HashDigest<SHA256>(value);
+
+                count += 1;
+            }
+        }
+
+        /// <inheritdoc/>
+        public override HashDigest<SHA256>? IndexBlockHash(Guid chainId, long index)
+        {
+            if (index < 0)
+            {
+                index += CountIndex(chainId);
+
+                if (index < 0)
+                {
+                    return null;
+                }
+            }
+
+            byte[] indexBytes = MySqlStoreBitConverter.GetBytes(index);
+
+            byte[] key = IndexKeyPrefix.Concat(indexBytes).ToArray();
+
+            var indices = Select<ChainModel, byte[]>(
+                ChainDbName,
+                "Key",
+                key);
+
+            byte[] bytes = null;
+            if (indices.Any())
+            {
+                foreach (var idx in indices)
+                {
+                    if (idx.Cf == chainId.ToString())
+                    {
+                        bytes = idx.Value;
+                    }
+                }
+            }
+
+            return bytes is null
+                ? (HashDigest<SHA256>?)null
+                : new HashDigest<SHA256>(bytes);
+        }
+
+        /// <inheritdoc/>
+        public override long AppendIndex(Guid chainId, HashDigest<SHA256> hash)
+        {
+            long index = CountIndex(chainId);
+            byte[] indexBytes = MySqlStoreBitConverter.GetBytes(index);
+            byte[] key = IndexKeyPrefix.Concat(indexBytes).ToArray();
+
+            QueryFactory db = new QueryFactory(new MySqlConnection(_connectionString), _compiler);
+
+            db.Statement(
+                "INSERT INTO Chain (`key`, `value`, `cf`, `prefix`) VALUES " +
+                "(@key1, @value1, @cf1, @prefix1), (@key2, @value2, @cf2, @prefix2) ON DUPLICATE KEY UPDATE " +
+                "`key` = VALUES(`key`), `value` = VALUES(`value`), `cf` = VALUES(`cf`), `prefix` = VALUES(`prefix`)",
```

Comment:
Why we don't use ORM?

Reply:
The reason for this raw sql was to use `ON DUPLICATE KEY UPDATE` clause which is currently not supported by Sqlkata. Regardless, I do think this makes the code look pretty messy so I'm thinking about building an ORM query for `INSERT INTO Chain (key, value, cf, prefix) VALUES (@key1, @value1, @cf1, @prefix1), (@key2, @value2, @cf2, @prefix2) ` and then possibly add a raw clause (`ON DUPLICATE KEY UPDATE key = VALUES(key), value = VALUES(value), cf = VALUES(cf), prefix = VALUES(prefix)`) at then end.